### PR TITLE
[Merged by Bors] - Sync wrong dbg assertion

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed

Running a beacon node I triggered a sync debug panic. And so finally the time to create tests for sync arrived. Fortunately, te bug was not in the sync algorithm itself but a wrong assertion

## Proposed Changes

- Split Range's impl from the BeaconChain via a trait. This is needed for testing. The TestingRig/Harness is way bigger than needed and does not provide the modification functionalities that are needed to test sync. I find this simpler, tho some could disagree.
- Add a regression test for sync that fails before the changes.
- Fix the wrong assertion.

### CODE REVIEW

**Review by AgeManning — APPROVED** [review]
Yeah nice. Good to get some sync tests going!

File: beacon_node/network/src/sync/range_sync/range.rs [hunk 1]
```diff
@@ -356,3 +361,257 @@ impl<T: BeaconChainTypes> RangeSync<T> {
...
+        let (finalized_peer, local_info, remote_info) = rig.finalized_peer();
+        range.add_peer(&mut rig.cx, local_info, finalized_peer, remote_info);
+        range.assert_state(RangeSyncType::Finalized);
+
+        // Sync should have requested a batch, grab the request
+        let _second_request = rig.grab_request(&finalized_peer);
+
+        // Fail the head chain by disconnecting the peer.
+        range.remove_peer(&mut rig.cx, &head_peer);
+        range.assert_state(RangeSyncType::Finalized);
```
Comment (dapplion) [inline beacon_node/network/src/sync/range_sync/range.rs:615]: I'm trying to understand what this assertion wanted to prove. Removing the head chain does not affect the sync state and it's already in Finalized type. What was the buggy behavior that this test fixed?

Comment (divagant-martian) [reply beacon_node/network/src/sync/range_sync/range.rs:615]: hey @dapplion, taking into account this was 3 years ago I most definitely do not remember the details. I restored the branch, if you want you can `gh pr checkout 2821` and revert the couple of assertions that I changed. Patch ahead

```diff
diff --git a/beacon_node/network/src/sync/range_sync/chain_collection.rs b/beacon_node/network/src/sync/range_sync/chain_collection.rs
index 4dc9c1d01..265c8b897 100644
--- a/beacon_node/network/src/sync/range_sync/chain_collection.rs
+++ b/beacon_node/network/src/sync/range_sync/chain_collection.rs
@@ -72,7 +72,7 @@ impl<T: BeaconChainTypes, C: BlockStorage> ChainCollection<T, C> {
             RangeSyncState::Finalized(ref syncing_id) => {
                 if syncing_id == id {
                     // the finalized chain that was syncing was removed
-                    debug_assert!(was_syncing && sync_type == RangeSyncType::Finalized);
+                    debug_assert!(was_syncing);
                     let syncing_head_ids: SmallVec<[u64; PARALLEL_HEAD_CHAINS]> = self
                         .head_chains
                         .iter()
@@ -86,7 +86,7 @@ impl<T: BeaconChainTypes, C: BlockStorage> ChainCollection<T, C> {
                     };
                 } else {
                     // we removed a head chain, or an stoped finalized chain
-                    debug_assert!(!was_syncing || sync_type != RangeSyncType::Finalized);
+                    debug_assert!(!was_syncing);
                 }
             }
             RangeSyncState::Head(ref mut syncing_head_ids) => {
@@ -507,4 +507,4 @@ impl<T: BeaconChainTypes, C: BlockStorage> ChainCollection<T, C> {
             }
         }
     }
-}
+}
\ No newline at end of file
```

and the test fails.

The specific assertion you ask, I don't know. However (at least back then) removing a peer might mean the state changes if it's the only peer on the chain that is syncing, for example. There might have been other reasons as well, but this is what comes to mind.

**Pull request closed without merging**